### ISSUE TITLE
connectivity issues with github. https silently fails but SSH works fine

### ISSUE THREAD

**Issue by 19379** [git-for-windows/git#390]
```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git clone https://github.com/19379/wsh_marc2003.git test
Cloning into 'test'...

User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote https://github.com/19379/wsh_marc2003.git
```

the cursor flashes for a few seconds then nothing. i tried to enable GIT_TRACE but i don't think it's helping...

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ export GIT_TRACE=1

User@HP-JMHBEJRPM4NE MINGW32 ~
$ git clone https://github.com/19379/wsh_marc2003.git test
19:22:04.306661 git.c:348               trace: built-in: git 'clone' 'https://github.com/19379/wsh_marc2003.git' 'test'
Cloning into 'test'...
19:22:05.753664 run-command.c:335       trace: run_command: 'git-remote-https' 'origin' 'https://github.com/19379/wsh_marc2003.git'
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote https://github.com/19379/wsh_marc2003.git
19:22:12.062698 git.c:348               trace: built-in: git 'ls-remote' 'https://github.com/19379/wsh_marc2003.git'
19:22:12.067698 run-command.c:335       trace: run_command: 'git-remote-https' 'https://github.com/19379/wsh_marc2003.git' 'https://github.com/19379/wsh_marc2003.git'
```

i can connect to github using curl...

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ curl -i https://github.com
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0HTTP/1.1 200 OK
Server: GitHub.com
Date: Mon, 14 Sep 2015 18:23:15 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Status: 200 OK
<snip>
```

and finally, i can clone with SSH just fine...
```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ export GIT_TRACE=0

User@HP-JMHBEJRPM4NE MINGW32 ~
$ git clone git@github.com:19379/wsh_marc2003.git test
Cloning into 'test'...
remote: Counting objects: 702, done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 702 (delta 14), reused 0 (delta 0), pack-reused 670
Receiving objects: 100% (702/702), 1.82 MiB | 387.00 KiB/s, done.
Resolving deltas: 100% (433/433), done.
Checking connectivity... done.
Checking out files: 100% (90/90), done.
```

given SSH works i don't have any real issue but i'm really curious about no errors being reported.

just to muddy the waters a little, the pc is configured to use a proxy but that's only via group policy/Internet Explorer and a PAC file. any application configured with no proxy can access/download anything they like. although browsing in IE is restricted, github is an allowed site and because i don't have any issues with curl or SSH i don't think it's that.

OS: windows 7 embedded 32bit. maybe an issue with not having a full fat OS? i'm totally clueless about this sort of thing!!

**Comment by dscho** [comment]
What Git for Windows version is this? And what does `echo $?` show after the failed commands? Does it work if you export the environment variable `CHARSET` (e.g. to `cp1252`)?

**Comment by 19379** [comment]
sorry i forgot to mention it was the latest 2.5.2 release 2.

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ export GIT_TRACE=1

User@HP-JMHBEJRPM4NE MINGW32 ~
$ git clone https://github.com/19379/wsh_marc2003.git test
20:18:30.101436 git.c:348               trace: built-in: git 'clone' 'https://github.com/19379/wsh_marc2003.git' 'test'
Cloning into 'test'...
20:18:31.696450 run-command.c:335       trace: run_command: 'git-remote-https' 'origin' 'https://github.com/19379/wsh_marc2003.git'

User@HP-JMHBEJRPM4NE MINGW32 ~
$ echo $?
128

User@HP-JMHBEJRPM4NE MINGW32 ~
$ export CHARSET=cp1252

User@HP-JMHBEJRPM4NE MINGW32 ~
$ git clone https://github.com/19379/wsh_marc2003.git test
20:23:09.864230 git.c:348               trace: built-in: git 'clone' 'https://github.com/19379/wsh_marc2003.git' 'test'
Cloning into 'test'...
20:23:11.487241 run-command.c:335       trace: run_command: 'git-remote-https' 'origin' 'https://github.com/19379/wsh_marc2003.git'

User@HP-JMHBEJRPM4NE MINGW32 ~
$ echo $?
128
```

**Comment by dscho** [comment]
128 most likely means: access violation...

**Comment by 19379** [comment]
well given it seems to be specific to only me and SSH is working, i'll close this. 

thanks for the prompt replies.

**Comment by dscho** [comment]
Indeed, this is what I got:

```sh
$ git version
git version 2.5.2.windows.2

$ git clone https://github.com/19379/wsh_marc2003.git test
Cloning into 'test'...
remote: Counting objects: 702, done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 702 (delta 14), reused 0 (delta 0Receiving objects: 100% (702/702)), pack-reused 670
Receiving objects: 100% (702/702), 1.82 MiB | 1.39 MiB/s, done.
Resolving deltas: 100% (433/433), done.
Checking connectivity... done.
```

However, I think it is a mistake to leave it at that. If you have that problem, so will others. This is your chance to contribute to the project.

**Comment by 19379** [comment]
i must confess i forgot to mention something else rather important. before using v2.5.2, i used this package of 1.9.5. this works fine over https on the same pc/internet connection.

https://github.com/sheabunge/GitPortable/releases/tag/v1.9.5-devtest.2

perhaps i could try some earlier 2.x builds and see if they work?

**Comment by dscho** [comment]
> 128 most likely means: access violation...

And since I cannot reproduce, I fear that you will have to hunt mostly on your own.

This is how I would start investigating, if I could reproduce:

1. install the [Git for Windows SDK](https://git-for-windows.github.io/#download-sdk)
2. verify that `git ls-remote https://github.com/19379/wsh_marc2003.git` still fails with exit code 128, even when running with the just-built Git.
3. modify the `remote-curl.c` file in `/usr/src/git/` by inserting `error("%s:%d: here", __FILE__, __LINE__); fflush(stderr);` lines in strategic places (e.g. at the beginning and the end of the `main()` function), probably also around the `curl_..._init()` calls in `http.c`.
4. run `make install` in `/usr/src/git/` and test again.

At some point, I would expect that the location of the crash could be pin-pointed to one particular line. This should be sufficient information to guess the true cause of the access violation.

**Comment by 19379** [comment]
right, i will try the above but it won't be for a few weeks. i don't have admin rights on the pc i use so i can't run any installer. i'm currently building my own pc but it won't be ready until the end of the month.

**Comment by dscho** [comment]
> i don't have admin rights on the pc i use so i can't run any installer.

Did you know that you can unpack the SDK installer with `.7z`? It is set up that way so that you *can* run that installer by unpacking it and then running the `setup-git-sdk.bat` script inside.

**Comment by 19379** [comment]
cool. i was simply put off by the filename. i'll give that a go tonight.

**Comment by dscho** [comment]
Very good. I will help you as much as I can, of course. If you run into troubles, just add a comment here.

**Comment by 19379** [comment]
can i ask how long this takes to build? there's no way i can use the pc i have connectivity issues with because it's a thin client with the weakest CPU known to man. even browsing the web is a fairly painful experience on it!! i do have access to a decent spec pc at my local library but time there is limited and given how big this project is, i can imagine it taking some time. it has a quad core i5 if that helps.

**Comment by dscho** [comment]
> i do have access to a decent spec pc at my local library

Hmm. Given that I cannot reproduce  this issue on any machine I have access to, I am rather doubtful that you can use that machine to investigate: simply switching the machine will make the symptom go away.

> can i ask how long this takes to build?

IIRC it will download a fair amount of data, like around 250MB. Depending on your machine, the build may take a long time; it takes around 5-20 minutes here, I guess, I never timed it precisely. But given your account, I imagine that it will take substantially longer on your machine to build (if you can clone the Git repository itself at all).

> also, will i get a choice of what to build when running the make file? i only want to build the 32bit portable install.

There is no choice, but it already builds what you need: the Git executables. You do not even need a portable install.

In fact, you only need a rebuilt `git-remote-https.exe` that is instrumented to show how far it gets before crashing.

**Comment by 19379** [comment]
>Hmm. Given that I cannot reproduce this issue on any machine I have access to, I am rather doubtful that you can use that machine to investigate: simply switching the machine will make the symptom go away.

i'm only intending to use the other pc to build. i'd obviously be testing it on the problematic pc.

>(if you can clone the Git repository itself at all).

i did notice that line near the end of the installer batch file. it's no problem for me to clone the repositories using SSH.

**Comment by dscho** [comment]
> i'm only intending to use the other pc to build. i'd obviously be testing it on the problematic pc.

That makes a *lot* of sense.

> it's no problem for me to clone the repositories using SSH

That should do, indeed.

**Comment by 19379** [comment]
i managed to build but still no output and i get the same `128` exit code. i put the changes i made here for you to double check...

https://github.com/19379/git/commit/78c874a6f9eef8c84ae8490f23ba85e4ddd8c2de

i just copied the new `git-remote-https.exe` like you said.

**Comment by dscho** [comment]
Okay, this establishes that it is *not* similar to the `CHARSET` problem as I thought earlier.

In addition to [this change](https://github.com/19379/git/commit/78c874a6f9eef8c84ae8490f23ba85e4ddd8c2de#commitcomment-13270391), you could also instrument the `start_command()` and `finish_command()` functions before rebuilding `git-fetch.exe` or `git-ls-remote.exe`, to figure out where exactly it exits with exit code 128.

I have a hunch that it is in [this line](https://github.com/git-for-windows/git/blob/0c51399ed604ac72935377eb1fba10c69f1bc01a/run-command.c#L248), but it could also be [this one](https://github.com/git-for-windows/git/blob/0c51399ed604ac72935377eb1fba10c69f1bc01a/run-command.c#L246).

The exit code *should* really be set [here](https://github.com/git-for-windows/git/blob/0c51399ed604ac72935377eb1fba10c69f1bc01a/compat/mingw.c#L2311).

**Comment by 19379** [comment]
^ i'm sorry, i don't understand what you want me to do??

also, i added a comment in reply to yours which you probably got a notification for...

https://github.com/19379/git/commit/78c874a6f9eef8c84ae8490f23ba85e4ddd8c2de#commitcomment-13271936

**Comment by dscho** [comment]
I meant to suggest instrumenting the code, like so: https://github.com/19379/git/pull/1 then rebuilding `git-ls-remote.exe` and executing that like `./git-ls-remote.exe https://github.com/19379/wsh_marc2003.git`.

**Comment by 19379** [comment]
new build, same result...

![git](https://cloud.githubusercontent.com/assets/12451008/9943140/806f586a-5d77-11e5-9152-2e6dea964688.png)

**Comment by dscho** [comment]
Did you copy the `git-ls-remote.exe` file you built to `/mingw32/libexec/git-core`? You should see at least *one line*, where `start_command()` tries to start `git-remote-https.exe`...

Ah, *slaps-his-head*, bummer! The `git-ls-remote.exe` is not the one you want to copy. You need to build `git.exe`, copy that into `/mingw32/bin/git.exe` and then run `/mingw32/bin/git ls-remote.exe ...`. I completely forgot that `git-ls-remote.exe` is *just* the Git wrapper, sorry!

**Comment by 19379** [comment]
see the huge bump in filesize!!!

![git](https://cloud.githubusercontent.com/assets/12451008/9946370/b414e4c4-5d8a-11e5-8672-4fcc58af7a3d.png)

as i forgot to include it on the screenshot, `echo $?` returns `128`

**Comment by dscho** [comment]
> see the huge bump in filesize!!!

That is to be expected, because you built it with debug symbols.

So now we're getting somewhere! `git-remote-https` was started. But it was not finished? And even worse: it seems that https://github.com/19379/git/commit/63ef029364b6e65090a14a77666d5564b8f6c4f4#diff-7577a5178f8cdc0f719e580577889f04R537 is hit, but not https://github.com/19379/git/commit/63ef029364b6e65090a14a77666d5564b8f6c4f4#diff-7577a5178f8cdc0f719e580577889f04R543?

Oh, but I forgot to put in `fflush(stderr);` after the `error()` calls. Under certain circumstances, `stderr` is actually not flushed, but cut off, in mintty if not flushed explicitly. Could you add those and try again?

**Comment by dscho** [comment]
Also, I am pretty certain that the `start_command()` function is called [here](https://github.com/19379/git/commit/63ef029364b6e65090a14a77666d5564b8f6c4f4#diff-7577a5178f8cdc0f719e580577889f04R543). You could instrument the surrounding code there, too, to figure out where things start to break down.

**Comment by 19379** [comment]
>Also, I am pretty certain that the start_command() function is called [here](https://github.com/19379/git/commit/63ef029364b6e65090a14a77666d5564b8f6c4f4#diff-7577a5178f8cdc0f719e580577889f04R543).

is that link right? it's pointing at a line which you've already modified. 

i've just built a fresh copy with [fflush added](https://github.com/19379/git/commit/d6233ab05276de4de77b679a2268199c704886b4) but won't be able to test until i get home later.

**Comment by dscho** [comment]
> > Also, I am pretty certain that the start_command() function is called here.
> 
> is that link right? it's pointing at a line which you've already modified.

Bah, copy-paste error. [This](https://github.com/19379/git/blob/d6233ab05276de4de77b679a2268199c704886b4/transport-helper.c#L132) is the location I meant.

**Comment by 19379** [comment]
is this ok?

```
error("Before start_command %s", cmd->argv[0]); fflush(stderr);
code = start_command(helper);
error("After start_command %s", cmd->argv[0]); fflush(stderr);
```

**Comment by 19379** [comment]
..and it's evidently not ok because it won't build!

```
106260@106260P MINGW32 /usr/src/git (master)
$ make install
GIT_VERSION = 2.5.2.windows.2.4.gd6233ab.dirty
    CC transport-helper.o
In file included from cache.h:4:0,
                 from transport-helper.c:1:
transport-helper.c: In function 'get_helper':
transport-helper.c:132:35: error: 'cmd' undeclared (first use in this function)
  error("Before start_command %s", cmd->argv[0]); fflush(stderr);
                                   ^
git-compat-util.h:439:27: note: in definition of macro 'error'
 #define error(...) (error(__VA_ARGS__), const_error())
                           ^
transport-helper.c:132:35: note: each undeclared identifier is reported only once for each function it appears in
  error("Before start_command %s", cmd->argv[0]); fflush(stderr);
                                   ^
git-compat-util.h:439:27: note: in definition of macro 'error'
 #define error(...) (error(__VA_ARGS__), const_error())
                           ^
git-compat-util.h:439:39: warning: left-hand operand of comma expression has no effect [-Wunused-value]
 #define error(...) (error(__VA_ARGS__), const_error())
                                       ^
transport-helper.c:132:2: note: in expansion of macro 'error'
  error("Before start_command %s", cmd->argv[0]); fflush(stderr);
  ^
git-compat-util.h:439:39: warning: left-hand operand of comma expression has no effect [-Wunused-value]
 #define error(...) (error(__VA_ARGS__), const_error())
                                       ^
transport-helper.c:134:2: note: in expansion of macro 'error'
  error("After start_command %s", cmd->argv[0]); fflush(stderr);
  ^
Makefile:1903: recipe for target 'transport-helper.o' failed
make: *** [transport-helper.o] Error 1
```

**Comment by dscho** [comment]
Yep, there is no `cmd` variable in that context. You probably need something like `helper->env_array.argv[0]`. Or something similar.

**Comment by 19379** [comment]
thanks, it's now building. i'm glad it did work because my time on the speedy pc at the library is about to run out.

**Comment by 19379** [comment]
right, here is the output of the latest build...

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: Started command git-remote-https
error: After start_command GIT_DIR=.git

User@HP-JMHBEJRPM4NE MINGW32 ~
$ echo $?
128
```

and then i had the bright idea of running it with SSH but it appears we've broken it!!

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote git@github.com:19379/jsplaylist-mod
error: Starting command ssh
error: Started command ssh
error: Finishing command ssh
error: compat/mingw.c:2314: got status 0
error: run-command.c:250: setting code 0
error: Finished command Segmentation fault

User@HP-JMHBEJRPM4NE MINGW32 ~
$ echo $?
139
```

and while clone appears to work, it also reports the same error..

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git clone git@github.com:19379/jsplaylist-mod
Cloning into 'jsplaylist-mod'...
error: Starting command ssh
error: Started command ssh
error: Starting command index-pack
error: Started command index-pack
remote: Counting objects: 38, done.
remote: Total 38 (delta 0), reused 0 (delta 0), pack-reused 38
Receiving objects: 100% (38/38), 221.12 KiB | 421.00 KiB/s, done.
Resolving deltas: 100% (6/6), done.
error: Finishing command index-pack
error: compat/mingw.c:2314: got status 0
error: run-command.c:250: setting code 0
error: Finished command index-pack (0)
error: Finishing command ssh
error: compat/mingw.c:2314: got status 0
error: run-command.c:250: setting code 0
error: Finished command Segmentation fault

User@HP-JMHBEJRPM4NE MINGW32 ~
$ echo $?
139
```

**Comment by 19379** [comment]
just writing to confirm that the build of `git.exe` created before your pull request works so that rules out problems with my build environment

```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote git@github.com:19379/jsplaylist-mod
869adbac09a7ae05201d9c8a825e5b22a801cae0        HEAD
869adbac09a7ae05201d9c8a825e5b22a801cae0        refs/heads/master

User@HP-JMHBEJRPM4NE MINGW32 ~
$ echo $?
0
```

i then tested with the build made from your pull request and that's where the errors with SSH come in.

**Comment by dscho** [comment]
The segmentation fault is probably from some careless print statement I introduced, e.g. when `cmd` was `free()`d before accessing it.

**Comment by 19379** [comment]
fixed... https://github.com/19379/git/commit/b1d1704ca9d261c717c3e54d29a0765d950f30d6

```
User@HP-JMHBEJRPM4NE MINGW32 ~/git (master)
$ git ls-remote git@github.com:19379/jsplaylist-mod
error: Starting command ssh
error: Started command ssh
error: Finishing command ssh
error: compat/mingw.c:2314: got status 0
error: run-command.c:250: setting code 0
error: Finished command ssh (0)
869adbac09a7ae05201d9c8a825e5b22a801cae0        HEAD
869adbac09a7ae05201d9c8a825e5b22a801cae0        refs/heads/master

User@HP-JMHBEJRPM4NE MINGW32 ~/git (master)
$ echo $?
0

User@HP-JMHBEJRPM4NE MINGW32 ~/git (master)
$ git ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: Started command git-remote-https
error: After start_command GIT_DIR=.git

User@HP-JMHBEJRPM4NE MINGW32 ~/git (master)
$ echo $?
128
```

...still not much to see in https mode.

**Comment by dscho** [comment]
... even if you force `git-remote-https.exe` to `exit(1);` as soon as it enters the `main()` function?

**Comment by 19379** [comment]
i edited `remote-curl.c` and put `exit(1);` as the first line inside the `main()` function. i then copied `git-remote-https.exe` to `mingw32\libexec\git-core`. i get code `128` as above.

i then renamed the file like i did before...

```
User@HP-2DU932P158BD MINGW32 ~
$ git ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: After start_command (null)
fatal: Unable to find remote helper for 'https'
```

**Comment by dscho** [comment]
Excellent. So now we know that

- the executable is getting launched
- but it does not load correctly.

Just a quick question to exclude other problems: the `git-remote-https.exe` you compiled is a 32-bit one, i.e. it was compiled in a 32-bit Git for Windows SDK, yes?

If so, then I suspect that there is some `.dll` search order problem. To make sure that it is something as mundane as that, could you make sure that the `post-install.bat` script has run if you install the portable Git manually?

**Comment by 19379** [comment]
as i was unsure about which batch file i should use to launch a console in the SDK folder, i've been using these bits copied from SDK installer batch file....

```
@FOR /F "delims=" %%D in ("%~dp0") do @set cwd=%%~fD
@set PATH=%cwd%\usr\bin;%PATH%
@set MSYSTEM=MINGW32
@start mintty -i /msys2.ico -t "Git SDK 32-bit" bash --login -i
```

also, i just ran the 32bit PortableGit installer in to a fresh folder and i can verify the `post-install.bat` file ran when it was done and then deleted itself. i get the same console messages reported above.

**Comment by dscho** [comment]
> also, i just ran the 32bit PortableGit installer in to a fresh folder and i can verify the post-install.bat file ran when it was done and then deleted itself. i get the same console messages reported above.

Thank you! This is a bit of a bummer, though, because it means that all the relevant `.dll` files were hard-linked into the correct folder (`libexec/git-core/`, so that the `.dll` search order -- which prefers the same directory as the executable over everything else -- plays in our favor).

Maybe `error("PATH: %s", getenv("PATH"));` just before https://github.com/git-for-windows/git/blob/0c51399ed604ac72935377eb1fba10c69f1bc01a/run-command.c#L479 might shed more light into the issue?

> as i was unsure about which batch file i should use to launch a console in the SDK folder, i've been using these bits copied from SDK installer batch file....
> 
> @FOR /F "delims=" %%D in ("%~dp0") do @set cwd=%%~fD
> @set PATH=%cwd%\usr\bin;%PATH%
> @set MSYSTEM=MINGW32
> @start mintty -i /msys2.ico -t "Git SDK 32-bit" bash --login -i

That is okay; you do not even need to bother with the first two lines, just the last two should be enough. But there should also be a `git-bash.exe` after everything was built; it is that executable that I use all the time to start the SDK (when it lives at the top-level directory of an SDK, that is).

**Comment by 19379** [comment]
```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: D:\Applications\PortableGit\mingw32/libexec/git-core;D:\Applications\PortableGit\home\bin;D:\Applications\PortableGit\mingw32\bin;D:\Applications\PortableGit\usr\local\bin;D:\Applications\PortableGit\usr\bin;D:\Applications\PortableGit\usr\bin;D:\Applications\PortableGit\mingw32\bin;D:\Applications\PortableGit\usr\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Program Files\Common Files\Trolltech\Qt-4.1.4;D:\Applications\PortableGit\usr\bin\vendor_perl;D:\Applications\PortableGit\usr\bin\core_perl
error: Started command git-remote-https
error: After start_command GIT_DIR=.git
```

**Comment by dscho** [comment]
> ;D:\Applications\PortableGit\home\bin;

Where does that come from? And: are there any `.dll` files in that directory?

**Comment by 19379** [comment]
i do have a `home` folder because i launch bash like a batch file like this...

```
set HOME=%~dp0home
set tmp=%~dp0tmp
start %~dp0git-bash.exe --cd-to-home
```

there is no `bin` folder inside it.

```D:\Applications\PortableGit\home>dir /b
.config
.ssh
.bash_history
.profile
.gitconfig
```

**Comment by dscho** [comment]
Hmm, okay.

Let me work a bit on your branch (although I notice that you did not push, but you should be able to merge my changes nevertheless).

**Comment by 19379** [comment]
i've just pushed that change.

**Comment by dscho** [comment]
Thanks!

**Comment by dscho** [comment]
With https://github.com/19379/git/pull/2, I get this:

```sh
$ ./git --exec-path=. ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: [...censored...]
error: Started command git-remote-https, returning 0
error: After start_command GIT_DIR=.git with code 0
error: code was 0
error: entering loop
error: remote-curl.c:961: here
error: http.c:497: here
error: http.c:499: here
error: http.c:355: here
error: http.c:357: here
error: got refspecs 0000000000000000
error: Finishing command git-remote-https
error: remote-curl.c:1043: here
error: compat/mingw.c:2314: got status 0
error: run-command.c:250: setting code 0
error: Finished command git-remote-https (0)
869adbac09a7ae05201d9c8a825e5b22a801cae0        HEAD
869adbac09a7ae05201d9c8a825e5b22a801cae0        refs/heads/master
```

**Comment by 19379** [comment]
```
User@HP-JMHBEJRPM4NE MINGW32 ~
$ git ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: D:\Applications\PortableGit\mingw32/libexec/git-core;D:\Applications\PortableGit\home\bin;D:\Applications\PortableGit\mingw32\bin;D:\Applications\PortableGit\usr\local\bin;D:\Applications\PortableGit\usr\bin;D:\Applications\PortableGit\usr\bin;D:\Applications\PortableGit\mingw32\bin;D:\Applications\PortableGit\usr\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Program Files\Common Files\Trolltech\Qt-4.1.4;D:\Applications\PortableGit\usr\bin\vendor_perl;D:\Applications\PortableGit\usr\bin\core_perl
error: Started command git-remote-https, returning 0
error: After start_command GIT_DIR=.git with code 0
error: code was 0
error: entering loop
```

**Comment by dscho** [comment]
Okay! So it happens in the loop.

**Comment by dscho** [comment]
See https://github.com/19379/git/pull/3

**Comment by dscho** [comment]
With this, I get these additional lines:

```
...
error: http.c:357: here
error: trying to receive a line
error: trying to receive a line
error: trying to receive a line
error: trying to receive a line
error: got refspecs 0000000000000000
...
```

**Comment by dscho** [comment]
Actually, I also set `debug = 1` in that PR and now the tail of my log looks like this:

```
error: http.c:357: here
Debug: Remote helper: <- fetch
Debug: Got cap fetch
error: trying to receive a line
Debug: Remote helper: Waiting...
Debug: Remote helper: <- option
Debug: Got cap option
error: trying to receive a line
Debug: Remote helper: Waiting...
Debug: Remote helper: <- push
Debug: Got cap push
error: trying to receive a line
Debug: Remote helper: Waiting...
Debug: Remote helper: <- check-connectivity
Debug: Got cap check-connectivity
error: trying to receive a line
Debug: Remote helper: Waiting...
Debug: Remote helper: <-
error: got refspecs 0000000000000000
Debug: Capabilities complete.
Debug: Remote helper: -> option progress false
Debug: Remote helper: Waiting...
Debug: Remote helper: <- ok
Debug: Remote helper: -> option verbosity 1
Debug: Remote helper: Waiting...
Debug: Remote helper: <- ok
Debug: Remote helper: Waiting...
Debug: Remote helper: <- @refs/heads/master HEAD
Debug: Remote helper: Waiting...
Debug: Remote helper: <- 869adbac09a7ae05201d9c8a825e5b22a801cae0 refs/heads/master
Debug: Remote helper: Waiting...
Debug: Remote helper: <-
Debug: Read ref listing.
Debug: Disconnecting.
error: Finishing command git-remote-https
error: remote-curl.c:1043: here
error: compat/mingw.c:2314: got status 0
error: run-command.c:250: setting code 0
error: Finished command git-remote-https (0)
869adbac09a7ae05201d9c8a825e5b22a801cae0        HEAD
869adbac09a7ae05201d9c8a825e5b22a801cae0        refs/heads/master
```

**Comment by 19379** [comment]
sorry, i ran the wrong command previously, i only did this...
```
git ls-remote https://github.com/19379/jsplaylist-mod
```

but now i don't know where i'm launching this from??

```
./git --exec-path=. ls-remote https://github.com/19379/jsplaylist-mod
```

i've only been putting new builds of `git.exe` inside `mingw32/bin` but i i try running the command from there, i get 

```
User@HP-JMHBEJRPM4NE MINGW32 /mingw32/bin
$ ./git --exec-path=. ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: <snip>
error: After start_command (null) with code -1
fatal: Unable to find remote helper for 'https'
```

anyway, i tried putting a new copy of `git.exe` inside `mingw32\libexec\git-core` and i got the same results as when i removed this from the edited post.

```
User@HP-JMHBEJRPM4NE MINGW32 /mingw32/libexec/git-core
$ ./git --exec-path=. ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: <snip>
\usr\bin\core_perl
error: Started command git-remote-https, returning 0
error: After start_command GIT_DIR=.git with code 0
error: code was 0
error: entering loop
```

just about to try the next one but if you could clarify what command i'm running from where??

**Comment by dscho** [comment]
Sorry, my call is from inside an SDK... You did the right thing to copy `git.exe` and `git-remote-https.exe` into `mingw32\libexec\git-core` and running it there.

**Comment by dscho** [comment]
FWIW I just saw that you can enable the built-in debug output by setting the environment variable `GIT_TRANSPORT_HELPER_DEBUG`.

**Comment by 19379** [comment]
```
User@HP-JMHBEJRPM4NE MINGW32 /mingw32/libexec/git-core
$ ./git --exec-path=. ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: <snip>
error: Started command git-remote-https, returning 0
error: After start_command GIT_DIR=.git with code 0
error: code was 0
Debug: Remote helper: -> capabilities
error: entering loop
error: trying to receive a line
Debug: Remote helper: Waiting...
Debug: Remote helper quit.
```

**Comment by dscho** [comment]
Oh, maybe the timeout is too small?

**Comment by dscho** [comment]
Try with this? https://github.com/19379/git/pull/4

**Comment by dscho** [comment]
Oh, maybe another sanity check?

```sh
$ ./git --exec-path=. remote-https
error: Starting command git-remote-https
error: PATH: [...the usual...]
error: Started command git-remote-https, returning 0
error: Finishing command git-remote-https
error: remote-curl.c:961: here
error: remote-curl: usage: git remote-curl <remote> [<url>]
error: compat/mingw.c:2314: got status 1
error: run-command.c:250: setting code 1
error: Finished command git-remote-https (1)
```

**Comment by 19379** [comment]
no change...

```
User@HP-JMHBEJRPM4NE MINGW32 /mingw32/libexec/git-core
$ ./git --exec-path=. ls-remote https://github.com/19379/jsplaylist-mod
error: Before start_command GIT_DIR=.git
error: Starting command git-remote-https
error: PATH: <snip>
error: Started command git-remote-https, returning 0
error: After start_command GIT_DIR=.git with code 0
error: code was 0
Debug: Remote helper: -> capabilities
error: entering loop
error: trying to receive a line
Debug: Remote helper: Waiting...
Debug: Remote helper quit.
```

and the sanity check...

```
User@HP-JMHBEJRPM4NE MINGW32 /mingw32/libexec/git-core
$ ./git --exec-path=. remote-https
error: Starting command git-remote-https
error: PATH: <snip>
error: Started command git-remote-https, returning 0
error: Finishing command git-remote-https
error: compat/mingw.c:2314: got status -1073741511
error: run-command.c:250: setting code 57
error: Finished command git-remote-https (57)
```

**Comment by dscho** [comment]
> error: compat/mingw.c:2314: got status -1073741511

Yep, I think that `remote-https` won't start. Can you start it directly, i.e. using `./git-remote-https.exe`?

```sh
$ ./git-remote-https.exe
error: remote-curl.c:961: here
error: remote-curl: usage: git remote-curl <remote> [<url>]
```

**Comment by dscho** [comment]
(Assuming that it does not start correctly, could you verify the hunch that it misses a `.dll` using [`depends.exe`](http://dependencywalker.com/)?)

**Comment by 19379** [comment]
```
User@HP-2DU932P158BD MINGW32 /mingw32/libexec/git-core
$ ./git-remote-https.exe

User@HP-2DU932P158BD MINGW32 /mingw32/libexec/git-core
$ echo $?
127
```

this looks potentially interesting.

https://cloud.githubusercontent.com/assets/12451008/9976938/08179e3c-5eed-11e5-93b4-496a5c949415.png

remember earlier when i said i was running windows 7 embedded.

**Comment by dscho** [comment]
I uploaded the picture here ;-)

![remote-https-dependencies](https://cloud.githubusercontent.com/assets/127790/9977001/2824e5b2-5ef8-11e5-869c-fafcdbcf4323.png)

The problem is the `.dll` files, like `libcurl-4.dll`, right? Are they not in `libexec/git-core/`?

For the record, this is what I get:

```sh
$ ls /mingw32/libexec/git-core/*.dll
/mingw32/libexec/git-core/edit.dll*
/mingw32/libexec/git-core/libatomic-1.dll*
/mingw32/libexec/git-core/libbz2-1.dll*
/mingw32/libexec/git-core/libcares-2.dll*
/mingw32/libexec/git-core/libcharset-1.dll*
/mingw32/libexec/git-core/libcurl-4.dll*
/mingw32/libexec/git-core/libeay32.dll*
/mingw32/libexec/git-core/libexpat-1.dll*
/mingw32/libexec/git-core/libffi-6.dll*
/mingw32/libexec/git-core/libgcc_s_dw2-1.dll*
/mingw32/libexec/git-core/libgmp-10.dll*
/mingw32/libexec/git-core/libgmpxx-4.dll*
/mingw32/libexec/git-core/libgnutls-30.dll*
/mingw32/libexec/git-core/libgomp-1.dll*
/mingw32/libexec/git-core/libgomp-plugin-host_nonshm-1.dll*
/mingw32/libexec/git-core/libhogweed-4-1.dll*
/mingw32/libexec/git-core/libiconv-2.dll*
/mingw32/libexec/git-core/libidn-11.dll*
/mingw32/libexec/git-core/libintl-8.dll*
/mingw32/libexec/git-core/libminizip-1.dll*
/mingw32/libexec/git-core/libnettle-6-1.dll*
/mingw32/libexec/git-core/libp11-kit-0.dll*
/mingw32/libexec/git-core/libpcre-1.dll*
/mingw32/libexec/git-core/librtmp-1.dll*
/mingw32/libexec/git-core/libssh2-1.dll*
/mingw32/libexec/git-core/libssp-0.dll*
/mingw32/libexec/git-core/libstdc++-6.dll*
/mingw32/libexec/git-core/libsystre-0.dll*
/mingw32/libexec/git-core/libtasn1-6.dll*
/mingw32/libexec/git-core/libtre-5.dll*
/mingw32/libexec/git-core/libvtv_stubs-0.dll*
/mingw32/libexec/git-core/libvtv-0.dll*
/mingw32/libexec/git-core/libwinpthread-1.dll*
/mingw32/libexec/git-core/ssleay32.dll*
/mingw32/libexec/git-core/tcl86.dll*
/mingw32/libexec/git-core/tk86.dll*
/mingw32/libexec/git-core/zlib1.dll*
```

after installing a fresh `PortableGit-2.5.3-32-bit.7z.exe` by double-clicking it...

**Comment by dscho** [comment]
(Note in particular the presence of `libcurl-4.dll`, `libiconv-2.dll`, `libintl-8.dll` and `libpcre-1.dll`...)

This is actually expected, as per [these lines in `13-copy-dlls.post`](https://github.com/git-for-windows/build-extra/blob/f7a89df8b979537a59d5758bc996ae345a72eb95/portable/root/etc/post-install/13-copy-dlls.post#L8-L19) which gets [called from `post-install.bat`](https://github.com/git-for-windows/build-extra/blob/master/post-install.bat#L41).

**Comment by 19379** [comment]
i had 0 dlls and now i know why.... it's because i'm using a FAT32 formatted thumb drive. installing to NTFS works fine. the link or copy does not copy when the link fails.

**Comment by dscho** [comment]
That's good to know, thanks! I will try to reproduce and if I can, I think I will be able to fix it.

Thanks for your persistence with this ticket!

**Comment by dscho** [comment]
Oops, I think I found it: the `echo` in [this line](https://github.com/git-for-windows/build-extra/blob/master/portable/root/etc/post-install/13-copy-dlls.post#L3) is a left-over from debugging and needs to go.

@19379 would you maybe able to test by unpacking the portable Git using 7-Zip, editing the `13-copy-dlls.post` file by deleting said `echo`, running `post-install.bat` and verifying that this fixes things on FAT (I have to go to bed now...)?

**Comment by 19379** [comment]
yes, that fixes it. i think we can close this now!

**Comment by dscho** [comment]
Not so fast ;-) The fix is not committed yet... Care to open a Pull Request? :grinning:

**Comment by 19379** [comment]
done.

**Comment by dscho** [comment]
Thanks!

**Comment by dscho** [comment]
@19379 no worries about the PR! It helped a lot because I only had to adjust it slightly to resolve the issue. Thank you so much!

**Comment by bviktor** [comment]
So what exactly is the purpose of a portable version if it requires manual "installation" steps? I used to be able to just extract the portable and deploy it everywhere, now I get things like this when starting Git bash:

~~~
mkdir: cannot change permissions of '/dev/shm': Permission denied
mkdir: cannot change permissions of '/dev/mqueue': Permission denied
'C:\\Windows\\System32\\drivers\\etc\\hosts' -> '/etc/hosts'
'C:\\Windows\\System32\\drivers\\etc\\protocol' -> '/etc/protocols'
'C:\\Windows\\System32\\drivers\\etc\\services' -> '/etc/services'
'C:\\Windows\\System32\\drivers\\etc\\networks' -> '/etc/networks'
~~~

If Git Portable now depends on its path or modifies files outside the Git folder, then it's not portable. 

If what the scripts does is the same for any path you put it in and doesn't modify files outside its own folder, then why isn't it done once, before packaging, instead of relying on users doing it?

TBH what the script does it pretty much gibberish to me so if someone cares to explain, that would be helpful.

I deploy Git Portable in a network share, so there's at least 3 paths where it's available: local path on the server (S:\tools\git2), network path on the server (\\whatever.domain\tools\git2), and local path when synced to the workstation (X:\whatever\path\they\use\git2). Besides, they're also pushed to Jenkins nodes with almost identical paths but different drive letters in some cases. These I could deploy by simply robocopy'ing them to any workstation and I obviously ain't gonna run scripts on every one of them. What do I do now?

**Comment by dscho** [comment]
> So what exactly is the purpose of a portable version if it requires manual "installation" steps?

I fail to see the relationship with "connectivity issues with github". Wrong ticket?

**Comment by bviktor** [comment]
No, I don't think so. The second half is about this "post-install" thingie.

**Comment by dscho** [comment]
> But I can open a new ticket if you wish.

Oh, come on. It is not like you are doing me a favor here. You are actually not doing yourself a favor, either, by dragging this out. How is your concern *possibly* related to connectivity issues with GitHub? It is not at all related to that, and you are intelligent enough to realize that.

And you are also smart enough to realize that a bug tracker has to be well organized in order to be effective. Discussing unrelated things in the same ticket (even if there might be some overlap in the code that is touched by those discussions) is doing everybody a disservice.

**Comment by bviktor** [comment]
A simple "yes, please" will do next time.

**Comment by dscho** [comment]
> A simple "yes, please" will do next time.

You still pretend that you are doing me a favor.

**Comment by bviktor** [comment]
Nope, I didn't consider my part a favor, the "please" part wasn't about "favors", only about human decency. 

I wasn't just randomly picking an issue to whine about a random thing. The release notes about 2.6.0 say "Portable Git's README is now clearer about the need to run post-install.bat when unpacking manually". This text links to pr 83, which is merely a few comments and it explains the changes by linking to **this** issue.

So I came to **this** issue to explain what kind of problems this change has caused. But I even said that I can open a new ticket if it's preferable.

Then you come and tell me it's not relevant (???), then 2 paragraphs of blah-blah about how I should treat you and your assaults to get a response out of your highness that's different from GTFO. Way to go, above and beyond.

**Comment by dscho** [comment]
You know, between replying to your rants, I actually also do real work. And I am getting a bit annoyed, I have to admit, that you still keep doing what you are doing, despite my clear indications that it is inappropriate.

**Comment by dscho** [comment]
(And yes, I also indicated what would be appropriate instead, so I was even constructive in my criticism.)
